### ISSUE TITLE
How to properly handle states?

### ISSUE THREAD

**Issue by lzhoucs** [dojo/compose#64]
We are in the process of converting a few ES6 classes into `ComposeFactory`. I ran into the following issue where I couldn't find a proper way to handle instance states.

Suppose I have the following sample class `FooBar` that I'd like to convert from:

``` typescript
interface Foo {
    getFoo(): string;
}

class FooBar implements Foo {
    protected log: string[];
    private count: number;
    constructor(log: string[]) {
        this.log = log;
        this.count = 0;
    };
    getFoo() {
        this.count++;
        this.log.push(`Have called ${this.count} times at ${new Date()}`);
        return 'foo';
    }
}
```

 When I attempt to to convert it into a `ComposeFactory` that creates `Foo` as shown below:

``` typescript
compose<Foo, Object>({
    getFoo(this: Foo) {
        this.count++;
        this.log.push(`Have called ${this.count} times at ${new Date()}`);
        return 'foo';
    }
});
```

It doesn't compile because `count` and `log` are not available in `Foo`, even though they can be set in a `Initializer` function. If I try to add them in the object literal:

``` typescript
compose<Foo, Object>({
    count: -1 as number,
    log: null as string[],
    getFoo(this: Foo) {
        this.count++;
        this.log.push(`Have called ${this.count} times at ${new Date()}`);
        return 'foo';
    }
})
```

I get the expected error `Object literal may only specify known properties`. There's a few workarounds, but none of them are ideal:
- Declare them in the interface `Foo`. This is not ideal since they are internal states of the instance.
- Remove generic types`<Foo, Object>`. I lose the benefit of type checking as well. I would like typescript to check to make sure I am implementing the right interface `Foo`.
- Use the original ES6 class `Foobar` as the base. It defeats the purpose of what I am doing - converting ES6 classes.

@maier49 [extracted a new state interface](https://github.com/maier49/store/blob/compose-refactory/src/store/createMemoryStore.ts#L148) along with [an external map](https://github.com/maier49/store/blob/compose-refactory/src/store/createMemoryStore.ts#L166) to maintain the states, in which case the state of the current instance has to be retrieved in every [function where it is needed](https://github.com/maier49/store/blob/compose-refactory/src/store/createMemoryStore.ts#L641).

Maybe there's a simpler way around.

**Comment by kitsonk** [comment]
ES6 classes don't have `private` or `protected` either.  They are TypeScript constructs and aren't present in the prototype of the class natively in ES6, and they are only enforced by convention.  There are future proposals for private properties in ECMAScript which differ greatly from the TypeScript conventions.

For values that are considered private and should only be accessed within the class, the pattern is to use a `WeakMap` with an object typed with any private properties (see also: dojo/meta#74).  For example:

``` typescript
import WeakMap from 'dojo-shim/WeakMap';

interface FooMap {
    count: number;
}

interface Foo {
    getFoo(): string;
}

const fooMap = new WeakMap<Foo, FooMap>();

const createFoo = compose<Foo, any>({
    getFoo(this: Foo) {
        fooMap.get(this).count++;
        return 'foo';
    }
}, (instance) => {
    fooMap.set(instance, {
        count: 0
    });
});
```

This pattern allows the sharing of information across instances of the same class, just like traditional OO allows for, but ensure that the information is truly private.

As far as the concept of `protected` I think we need to understand the use case, especially when we want to focus on a mixin concept, there really shouldn't be "sharing" of information with descendants and that either the information is _public_ or it is _private_.  @maier49 and I discussed that there might be the need to specify the shape of a mixin that another mixin requires, but I don't know if we want to go as far as creating "mixin only" interfaces unless as there is a compelling use case.

**Comment by lzhoucs** [comment]
Thanks for the quick and detailed response @kitsonk. Sorry I mixed the concept of ES6 classes and TS classes since they are so similar. And since we are doing TypeScript, what I meant by ES6 classes was TS classes. I saw a couple of other places referring to TS classes as ES6 classes as well.

I didn't realize what @maier49 did with the map is the recommended pattern to use. I will go with this pattern as well.
